pkg/app/piped/appconfigreporter/appconfigreporter.go

```diff
@@ -0,0 +1,366 @@
+	headCommits := make(map[string]string, len(r.gitRepos))
+	for repoID, repo := range r.gitRepos {
+		var headCommit git.Commit
+		headCommit, err = repo.GetLatestCommit(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get the latest commit of %s: %w", repoID, err)
+		}
+		// Skip if the head commit is already scanned.
+		lastScannedCommit, ok := r.lastScannedCommits[repoID]
+		if ok && headCommit.Hash == lastScannedCommit {
+			continue
+		}
+		var as []*model.ApplicationInfo
+		as, err = r.findRegisteredApps(ctx, repoID, repo, lastScannedCommit, headCommit.Hash, registeredAppPaths)
+		if err != nil {
+			return err
+		}
+		apps = append(apps, as...)
+		headCommits[repoID] = headCommit.Hash
+	}
+	defer func() {
+		if err == nil {
+			for repoID, hash := range headCommits {
+				r.lastScannedCommits[repoID] = hash
+			}
+		}
+	}()
+	if len(apps) == 0 {
+		return
+	}
+
+	_, err = r.apiClient.UpdateApplicationConfigurations(
+		ctx,
+		&pipedservice.UpdateApplicationConfigurationsRequest{
+			Applications: apps,
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to update application configurations: %w", err)
+	}
+	return
+}
+
+// findRegisteredApps finds out registered application info in the given git repository.
+func (r *Reporter) findRegisteredApps(ctx context.Context, repoID string, repo gitRepo, lastScannedCommit, headCommitHash string, registeredAppPaths map[string]struct{}) ([]*model.ApplicationInfo, error) {
+	if lastScannedCommit == "" {
+		return r.scanAllFiles(ctx, repo.GetPath(), repoID, registeredAppPaths, true)
+	}
+
+	files, err := repo.ChangedFiles(ctx, lastScannedCommit, headCommitHash)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get files those were touched between two commits: %w", err)
+	}
+	if len(files) == 0 {
+		// The case where all changes have been fully reverted.
+		return []*model.ApplicationInfo{}, nil
+	}
+	apps := make([]*model.ApplicationInfo, 0)
+	for _, filename := range files {
+		if shouldSkip(repoID, filename, registeredAppPaths, true) {
+			continue
+		}
+		appInfo, err := r.readApplicationInfo(ctx, filepath.Join(repo.GetPath(), filename))
+		if err != nil {
+			r.logger.Error("failed to read application info",
+				zap.String("repo-id", repoID),
+				zap.String("config-file-path", filename),
+				zap.Error(err),
+			)
+			continue
+		}
+		apps = append(apps, appInfo)
+	}
+	return apps, nil
+}
+
+// updateUnregisteredApps sends all unregistered application configurations to the control-plane.
+func (r *Reporter) updateUnregisteredApps(ctx context.Context, registeredAppPaths map[string]struct{}) error {
+	apps := make([]*model.ApplicationInfo, 0)
+	for repoID, repo := range r.gitRepos {
+		as, err := r.findUnregisteredApps(ctx, repo.GetPath(), repoID, registeredAppPaths)
+		if err != nil {
+			return err
+		}
+		r.logger.Info(fmt.Sprintf("found out %d unregistered applications in repository %s", len(as), repoID))
+		apps = append(apps, as...)
+	}
+	if len(apps) == 0 {
+		return nil
```

Comment:
Do you think that we should send this empty list to the control plane to delete the cache?

Reply:
Or we can have a simple mark in this Reporter to avoid sending continuous this kind of request.

Reply:
To avoid sending meaningless requests simply, I applied this way. Cache in Control-plane gets periodically removed, so I thought it's fine. But it's worth considering.